node/src/components/chainspec_loader.rs

```diff
@@ -208,8 +214,24 @@ where
         event: Self::Event,
     ) -> Effects<Self::Event> {
         match event {
-            Event::Request(ChainspecLoaderRequest::GetChainspecInfo(req)) => {
-                req.respond(self.clone().into()).ignore()
+            Event::Request(ChainspecLoaderRequest::GetChainspecInfo(responder)) => {
+                responder.respond(self.clone().into()).ignore()
+            }
+            Event::Request(ChainspecLoaderRequest::NextUpgradeActivationPoint(responder)) => {
+                let root_dir = self.root_dir.clone();
+                let current_version = self.chainspec.protocol_config.version.clone();
+                async move {
+                    let maybe_upgrade_activation_point = task::spawn_blocking(move || {
+                        next_activation_point(root_dir, current_version)
+                    })
+                    .await
+                    .unwrap_or_else(|error| {
+                        warn!(%error, "failed to join tokio task");
+                        None
+                    });
```

Comment:
If this keeps happening repeatedly then a node might miss the upgrade and create a fork in the network. We might want to reconsider the severity of this.